File: Python/297. Serialize and Deserialize Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Codec:

    def __init__(self):
        self.str = ""

    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        if root == None:
            return "#"
        res = str(root.val) + "#"
        res += self.serialize(root.left)
        res += self.serialize(root.right)
        return res

    def des(self):
        num = 0
        first = self.str.find("#")
        if first == 0:
            self.str = self.str[1:]
            return None
        root = TreeNode(int(self.str[0:first]))
        self.str = self.str[first + 1:]
        root.left = self.deserialize(self.str)
        root.right = self.deserialize(self.str)
        return root

    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        self.str = data
        return self.des()


# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))

File: 2023/297. Serialize and Deserialize Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        string res;
        queue<TreeNode*> queue;
        queue.push(root);
        while (!queue.empty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode* curr = queue.front();
                queue.pop();
                if (curr == NULL) {
                    res += "##";
                } else {
                    res += to_string(curr->val) + "#";
                    queue.push(curr->left);
                    queue.push(curr->right);
                }
            }
        }
        return res;
    }

    int getNext(string& data, int& index) {
        int val = 0, flag = 1;
        if (data[index] == '-') {
            flag = -1;
            index++;
        }
        while (data[index] != '#') {
            val = val * 10 + data[index] - '0';
            index++;
        }
        index++;
        return val * flag;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if (data == "##") {
            return NULL;
        }
        int index = 0;
        queue<TreeNode*> queue;
        TreeNode* root = new TreeNode(getNext(data, index));
        queue.push(root);
        while (!queue.empty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode* parent = queue.front();
                queue.pop();
                if (data[index] == '#') {
                    index += 2;
                } else {
                    TreeNode* left = new TreeNode(getNext(data, index));
                    parent->left = left;
                    queue.push(left);
                }
                if (data[index] == '#') {
                    index += 2;
                } else {
                    TreeNode* right = new TreeNode(getNext(data, index));
                    parent->right = right;
                    queue.push(right);
                }
            }
        }
        return root;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));